core/src/main/java/io/grpc/ClientInterceptors.java

```diff
@@ -204,4 +212,190 @@ public final void start(Listener<RespT> responseListener, Metadata headers) {
+      private ReqT requestMessage;
+      private boolean compressionEnabled;
+      private final Queue<AttemptListener> attemptListeners =
+              new ConcurrentLinkedQueue<AttemptListener>();
+      private volatile AttemptListener latestResponse;
+      private final RetryPolicy retryPolicy;
+
+      RetryingCall(MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel channel,
+                   RetryPolicy retryPolicy) {
+        this.method = method;
+        this.callOptions = callOptions;
+        this.channel = channel;
+        this.retryPolicy = retryPolicy;
+      }
+
+      @Override
+      public void start(Listener<RespT> listener, Metadata headers) {
+        checkState(attemptListeners.isEmpty());
+        checkState(responseListener == null);
+        checkState(requestHeaders == null);
+        responseListener = listener;
+        requestHeaders = headers;
+        ClientCall<ReqT, RespT> firstCall = channel.newCall(method, callOptions);
+        AttemptListener attemptListener = new AttemptListener(firstCall);
+        attemptListeners.add(attemptListener);
+        firstCall.start(attemptListener, headers);
+      }
+
+      @Override
+      public void request(int numMessages) {
+        checkState(!attemptListeners.isEmpty());
+        lastCall().request(numMessages);
+      }
+
+      @Override
+      public void cancel() {
+        for (AttemptListener attempt : attemptListeners) {
+          attempt.call.cancel();
+        }
+      }
+
+      @Override
+      public void halfClose() {
+        checkState(!attemptListeners.isEmpty());
+        lastCall().halfClose();
+      }
+
+      @Override
+      public void sendMessage(ReqT message) {
+        checkState(requestMessage == null);
+        checkState(!attemptListeners.isEmpty());
+        requestMessage = message;
+        lastCall().sendMessage(message);
+      }
+
+      @Override
+      public boolean isReady() {
+        checkState(!attemptListeners.isEmpty());
+        return lastCall().isReady();
+      }
+
+      @Override
+      public void setMessageCompression(boolean enabled) {
+        compressionEnabled = enabled;
+        lastCall().setMessageCompression(enabled);
+      }
+
+      private void maybeRetry(AttemptListener attempt) {
+        if (!retryPolicy.isRetryable(method, callOptions)) {
+          return;
+        }
+
+        if (attempt.responseStatus.isOk()) {
+          useResponse(attempt);
+          return;
+        }
+
+        Status.Code code = attempt.responseStatus.getCode();
+        if (code == Status.Code.CANCELLED || code == Status.Code.DEADLINE_EXCEEDED) {
+          AttemptListener latest = latestResponse;
+          if (latest != null) {
+            useResponse(latest);
+          } else {
+            useResponse(attempt);
+          }
+          return;
+        }
+
+        // Backoff before retry
+        try {
+          Thread.sleep(retryPolicy.getNextBackoffMillis());
```

Comment:
Definitely don't want to sleep here. Instead, you'd need to schedule the subsequent attempt using a `ScheduledExecutorService`.

I can't recall exactly what the threading guarantees are, though, for calling the listener. But since you're only invoking methods on a new call object (so they are all invoked serially), I think you are fine. Internally, the call has a `callExecutor` to serialize invocations to the listener.